core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,710 @@
+                    'tokens': list(str). Token list of program.
+                    'k_gram_hashes': list(int). K-gram hash values of program.
+                    'fingerprint': list(int). Extracted fingerprint of program.
+                } for each program in training_data.
+            }
+    """
+    window_size = T - K + 1
+    for program_id in training_data:
+        training_data[program_id]['fingerprint'] = (
+            winnowing.get_fingerprint_from_hashes(
+                training_data[program_id]['k_gram_hashes'], window_size))
+    return training_data
+
+
+def calc_jaccard_index(multiset_a, multiset_b):
+    """Calculate jaccard's coefficient for two multisets mutliset_a
+    and multiset_b.
+
+    Jaccard index of two set is equal to:
+        (no. of elements in intersection of two multisets)
+        _____________________________________________
+        (no. of elements in union of two multisets)
+
+    Note: intersection and union of two multisets is similar to union-all and
+    intersect-all operations in SQL.
+
+    Args:
+        multiset_a: list(int). First set.
+        multiset_b: list(int). Second set.
+
+    Returns:
+        float. Jaccard index of two sets.
+
+    """
+    small_set = (
+        multiset_a[:] if len(multiset_a) < len(multiset_b) else multiset_b[:])
+    union_set = (
+        multiset_b[:] if len(multiset_a) < len(multiset_b) else multiset_a[:])
+    small_set = sorted(small_set)
+    union_set = sorted(union_set)
+    index = 0
+    extra_elements = []
+    for elem in small_set:
+        while index < len(union_set) and elem > union_set[index]:
+            index += 1
+        if index >= len(union_set) or elem < union_set[index]:
+            extra_elements.append(elem)
+        elif elem == union_set[index]:
+            index += 1
+
+    union_set.extend(extra_elements)
+    if union_set == []:
+        return 0
+
+    multiset_a = sorted(multiset_a[:])
+    multiset_b = sorted(multiset_b[:])
+    index = 0
+    intersection_set = []
+    for elem in multiset_a:
+        while index < len(multiset_b) and elem > multiset_b[index]:
+            index += 1
+        if index < len(multiset_b) and elem == multiset_b[index]:
+            index += 1
+            intersection_set.append(elem)
+
+    coeff = float(len(intersection_set)) / len(union_set)
+    return coeff
+
+
+def get_program_similarity(fingerprint_a, fingerprint_b):
+    """Find similarity between fingerprint of two programs.
+
+    Args:
+        fingerprint_a: list((int, int)). Fingerprint of first data point.
+        fingerprint_b: list((int, int)). Fingerprint of second data point.
+
+    Returns:
+        float. Similarity between first and second program.
+    """
+    multiset_a = [h for (h, _) in fingerprint_a]
+    multiset_b = [h for (h, _) in fingerprint_b]
+    return calc_jaccard_index(multiset_a, multiset_b)
+
+
+def add_top_similars(training_data, top):
```

Comment:
similars? perhaps use "K" instead of "top" and say `add_k_nearest_neighbours`. K sounds more familiar.